server/src/main/java/org/candlepin/async/tasks/RefreshPoolsForProductJob.java

```diff
@@ -0,0 +1,147 @@
+import org.candlepin.pinsetter.core.model.JobStatus;
+import org.candlepin.service.OwnerServiceAdapter;
+import org.candlepin.service.SubscriptionServiceAdapter;
+import org.candlepin.util.Util;
+
+import com.google.inject.Inject;
+
+import org.apache.log4j.MDC;
+
+import java.util.Objects;
+
+
+/**
+ * RefreshPoolsForProductJob
+ */
+public class RefreshPoolsForProductJob implements AsyncJob {
+
+    private PoolManager poolManager;
+    private ProductCurator productCurator;
+    private SubscriptionServiceAdapter subAdapter;
+    private OwnerServiceAdapter ownerAdapter;
+
+    public static final String LAZY_REGEN = "lazy_regen";
+
+    public static final String JOB_KEY = "REFRESH_POOLS_FOR_PRODUCT";
+    private static final String JOB_NAME = "refresh_pools_for_product";
+
+    @Inject
+    public RefreshPoolsForProductJob(
+        final ProductCurator productCurator,
+        final PoolManager poolManager,
+        final SubscriptionServiceAdapter subAdapter,
+        final OwnerServiceAdapter ownerAdapter) {
+        this.poolManager = Objects.requireNonNull(poolManager);
+        this.productCurator = Objects.requireNonNull(productCurator);
+        this.subAdapter = Objects.requireNonNull(subAdapter);
+        this.ownerAdapter = Objects.requireNonNull(ownerAdapter);
+    }
+
+    @Override
+    public String execute(JobExecutionContext context) {
+        final JobArguments args = context.getJobArguments();
+
+        final String productUuid = args.getAsString(JobStatus.TARGET_ID);
+        final Boolean lazy = args.getAsBoolean(LAZY_REGEN);
+        final StringBuilder result = new StringBuilder();
+
+        final Product product = this.productCurator.get(productUuid);
+
+        if (product != null) {
+            final Refresher refresher = poolManager.getRefresher(this.subAdapter, this.ownerAdapter, lazy);
+
+            refresher.add(product);
+            refresher.run();
+
+            result.append("Pools refreshed for product: ")
+                .append(productUuid)
+                .append("\n");
+        }
+        else {
+            result.append("Unable to refresh pools for product \"")
+                .append(productUuid)
+                .append("\": Could not find a product with the specified UUID");
+        }
+
+        return result.toString();
+    }
+
+    /**
+     * Creates a JobConfig configured to execute the export job. Callers may further manipulate
+     * the JobConfig as necessary before queuing it.
+     *
+     * @return
+     *  a JobConfig instance configured to execute the export job
+     */
+    public static JobConfig createJobConfig(final Product product, final boolean lazy) {
+        return new RefreshPoolsForProductConfig(product, lazy);
+    }
+
+    /**
+     * Job configuration object for the export job
+     */
+    public static class RefreshPoolsForProductConfig extends JobConfig {
+
+        RefreshPoolsForProductConfig(final Product product, final boolean lazy) {
```

Comment:
This somewhat violates the fluent-style design we're going with for the per-job configuration objects. Ideally this is a default constructor and, (optionally,) provides setters for the required parameters. Take a look at the updated ExportJob and the PR for the ImportJob for examples.

Additionally, there's no validation on the inbound Product object before we go using it down on line 116.

Reply:
Does it make sense to provide mandatory parameters through setters? It causes quite strong temporal coupling as it makes it easy to forget or accidentally remove required parameter and the only way to know that it belongs there is to look at the implementation.

Reply:
Yeah, that's a fair point that's been discussed a couple times. The options were basically parameter soup methods as we've had in the past, or the fluent-style syntax which conveys less information about requirements, but is easier to work with in code.

We're (currently) opting for the fluent-style due to the frequency of random usage of the jobs, but I'm definitely open to a 3rd option that moves us away from passing in a pile of arguments into a single method/constructor, but also is good at signaling its requirements/usage to those not immediately familiar with it.